server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -268,6 +269,7 @@ Runnable getGlobalCheckpointSyncer() {
 
     private final AtomicLong lastSearcherAccess = new AtomicLong();
     private final AtomicReference<Translog.Location> pendingRefreshLocation = new AtomicReference<>();
+    private volatile boolean useRetentionLeasesInPeerRecovery;
```

Comment:
I tried with https://github.com/elastic/elasticsearch/pull/49448/commits/a544d77d4d58d73cf441877911a1a0768a371f0d. However, I prefer the current approach because: 

- `hasAllPeerRecoveryRetentionLeases` is slightly different from `useRetentionLeasesInPeerRecovery` since the former becomes true after we have created PRRLs for every **tracked** copy. We need to be more careful when changing this flag on replicas since we might miss creating PRRL if we set this flag too soon.

- `hasAllPeerRecoveryRetentionLeases` can be updated asynchronously, hence we need to introduce a new callback

- We will break some invariants if we decide to update the `routingTable` on replicas. I think we should keep these invariants.

I am fine if we decide to go with your suggestion.